### ISSUE TITLE
Feature Request: Add Modular DRL Model Architecture (DQN Ready)

### ISSUE THREAD

**Issue by ProValarous** [UHUMALAB/PPAGE#11]
## Overview

This issue proposes adding **Deep Reinforcement Learning (DRL) model definitions** to the project so agents can later use deep networks instead of tabular methods.

The goal is **not to add training code or RL algorithms**, but to define **clean, reusable neural network models** that can be easily plugged into agents in the future.

This will make it easier for new contributors to experiment with DRL while keeping the codebase clean and maintainable.

---

## Why This Is Needed

The Predator-Prey Gridworld environment is ideal for studying coordination and learning in multi-agent systems. However, as state spaces grow, tabular methods do not scale well.

Adding modular DRL models allows:

* Scaling to larger environments
* Learning useful representations from observations
* Fair comparison between tabular and deep approaches
* Future extension to algorithms like DQN, IQL, VDN, or QMIX

This issue focuses only on **model architecture**, not learning logic.

---

### Included

* Neural network models compatible with DQN-style methods
* Modular and reusable design
* Plug-and-play models that agents can use without modification
* Configuration-based architecture and hyperparameters
* Initial implementation using **MLP (fully connected) networks**

### Not Included

* Training loops
* Replay buffers
* Loss functions or optimizers
* Any RL algorithm implementation

Keeping this scope limited is intentional.

---

### 1. Plug-and-Play Models

Models should be independent of:

* The environment
* The agent logic
* The training algorithm

Agents should be able to use a model by simply calling it with an observation.

Example:

```python
q_values = model(observation)
```

No exploration logic, rewards, or environment details should exist inside the model.

---

### 2. Modular Structure

Recommended folder structure:

```
models/
  base/
    agent_model.py
  encoders/
    mlp_encoder.py
  heads/
    q_head.py
  configs/
    dqn_mlp.yaml
```

* **Encoders** process observations into embeddings
* **Heads** output Q-values
* **AgentModel** combines encoder and head

This makes it easy to extend later without breaking existing code.

---

### 3. Start Simple: MLP Only

The first implementation should use:

* MLP encoder for vector observations
* DQN-style Q-value head

More complex architectures (CNNs, GNNs, RNNs) can be added later once the basics are stable.

---

### 4. Configuration Files

All model parameters and hyperparameters should be stored in config files, not hardcoded.

Examples:

* Number of layers
* Hidden dimensions
* Activation functions

This improves readability, reproducibility, and experimentation.

---

### 5. Parameter Sharing Support

The design should allow **one shared model across multiple agents**.

If agent identity is needed, it should be handled through optional embeddings, not separate networks.

---

### 6. Open Source Best Practices

To keep the project contributor friendly:

* Use clear docstrings and type hints
* Add brief documentation for each module
* Keep model code free from environment assumptions
* Follow consistent naming and formatting
* Support deterministic initialization where possible

---

## Expected Deliverables

* Base `AgentModel` interface
* MLP encoder implementation
* DQN-compatible Q-value head
* Example config file
* Minimal documentation explaining how to use and extend the models

No training or learning code is expected in this issue.

---

## Future Work (Out of Scope)

* Training pipelines
* Multi-agent learning algorithms
* Centralized critics or mixing networks
* Advanced model architectures

These should be addressed in separate issues after this foundation is complete.

---

## Final Note for New Contributors

If you are new to the project:

* You do not need deep RL experience to work on this
* Focus on clean PyTorch code and modular design
* Ask questions if any interface or design choice is unclear

**Comment by ProValarous** [comment]
It is Done!
